README.md
#Autotask

autotask is a **django-application** for handling asynchronous tasks without the need to install, configure and supervise additional processes (like celery, redis or rabbitmq). autotask is aimed for applications where asynchronous tasks happend occasionally and the installation, configuration and monitoring of an additionally technology stack seems to be to much overhead.

##Requirements

- Python >= 3.3, 2.7
- Django >= 1.8
- Database: PostgreSQL


##Installation

Download and install using pip:

    pip install autotask

Register *autotask* as application in *settings.py*:

    # Application definition
    INSTALLED_APPS = [
        ...
        'autotask',
    ]

Run migrations to install the database-table used by *autotask*:

    $ python manage.py migrate


##Usage

Activate *autotask* in *settings.py* ::

    AUTOTASK_IS_ACTIVE = True

**Notes:**

- Don't activate *autotask* before running ``python manage.py migrate``. Otherwise *autotask* will try to access an undefined database-table.
- Don't run test with *autotask* activated. This will break tests because of an atexit-handler.
- If *autotask* has not shutdown properly (because of a *kill 9* or some strange crash) a supervisor-marker may stay in the database preventing to start autotask the next time. In this case just restart the django process.


**autotask** offers three decorators for handling asynchronous task:

    from autotask.tasks import (
        delayed_task,
        periodic_task,
        cron_task,
    )


###delayed_task

    @delayed_task(delay=0, retries=0, ttl=300)
    def some_function(*args, **kwargs):
        ...

A call to a function decorated by **@delayed_task()** will return immediately. The function itself will get executed later in another process. The decorator takes the following optional arguments:

- **delay**: time in seconds to wait at least befor the function gets executed. Defaults to 0 (as soon as possible).
- **entries**: Number of retries to execute a function in case of a failure. Defaults to 0 (no retries).
- **ttl**: time to live. After running a function the result will be stored at least for this time. Defaults to 300 seconds.

The decorated function returns an object with the following attributes:

**ready**: True if the task has been executed or False in case the task is still waiting for execution.

**status**: Can have the following values (which can be imported from autotask.task):

    from autotask.task import (
        WAITING,
        RUNNING,
        DONE,
        ERROR
    )

    - WAITING: tasks waits for execution
    - RUNNING: task gets executed right now
    - DONE: task has been executed
    - ERROR: an error has occured during the execution

**result**: the result of the executed task


###periodic_task

    @periodic_task(seconds=3600, start_now=False)
    def some_function(*args, **kwargs):
        ...

A function decorated by **@periodic_task()** should not get called but has to be imported when django starts up to execute the decorator. This will register the function to get executed periodically. The decorator takes the following optional arguments:

**seconds**: time in seconds to wait before executing the function again. Defaults to 3600 (an hour).

**start_now**: a boolean value: if *True* execute as soon as possible and then periodically. If *False* wait for the given number of seconds before running periodically. Defaults to *False*.


###cron_task

    @cron_task(minutes=None, hours=None, dow=None,
               months=None, dom=None, crontab=None)
    def some_function(*args, **kwargs):
        ...

A function decorated by **@cron_task()** should not get called but has to be imported when django starts up to execute the decorator. This will register the function to get executed according to the crontab-arguments. These arguments can  be given as python sequences or as a crontab-string.

**minutes**: list of minutes during an hour when the task should run. Valid entries are integers in the range 0-59. Defaults to None which is the same as '*' in a crontab, meaning that the tasks gets executed every minute.

**hours**: list of hours during a day when the task should run. Valid entries are integers in the range 0-23. Defaults to None which is the same as '*' in a crontab, meaning that the tasks gets executed every hour.

**dow**: days of week. A list of integers from 0 to 6 with Monday as 0. The task runs only on the given weekdays. Defaults to None which is the same as '*' in a crontab, meaning that the tasks gets executed every day of the week.

**months**: list of month during a year when the task should run. Valid entries are integers in the range 1-12. Defaults to None which is the same as '*' in a crontab, meaning that the tasks gets executed every month.

**dom**: list of days in an month the task should run. Valid entries are integers in the range 1-31. Defaults to None which is the same as '*' in a crontab, meaning that the tasks gets executed every day.

If neither **dom** nor **dow** are given, then the task should run every day of a month. If one of both are set, then the given restrictions apply. If both are set, then the allowed days are complement each other.

**crontab**: a string representing a valid crontab. See: [https://en.wikipedia.org/wiki/Cron#CRON_expression](https://en.wikipedia.org/wiki/Cron#CRON_expression) with the restriction that only integers and the special signs (* , -) are allowed. Some examples:

    - * * * * *: runs every minute
    - 15,30 7 * * *: runs every day at 7:15 and 7:30
    - * 9 0 4,7 10-15: runs at 9:00 every monday and from the 10th to the 15th of a month but only in April and July.

If the argument **crontab** is given all other arguments are ignored.



##Settings

All settings are optional and preset with default values. To override these defaults redefine them in the *settings.py* file.

**``AUTOTASK_IS_ACTIVE``**: Boolean. If *True* autotask will start a worker-process to handle the decorated tasks. Defaults to *False* (for easiers installation).

**``AUTOTASK_WORKERS``**: Integer. Number of worker-processes to start. Defaults to 1. (new in version 0.6)

**``AUTOTASK_WORKER_EXECUTABLE``**: String. Path to the executable for *manage.py <command>*. Must be absolute or relative to the working directory defined by BASE_DIR in the *settings.py* file. Defaults to "python" without a leading path.

**``AUTOTASK_WORKER_MONITOR_INTERVALL``**: Integer. Time in seconds for autotask to check whether the worker process is alive. Defaults to 5.

**``AUTOTASK_HANDLE_TASK_IDLE_TIME``**: Integer. Time in seconds to sleep on idle times. After processing a task autotask checks for the next task and executes it without delay if its scheduled for the current time. If no scheduled task is found autotasks sleeps for the given time in seconds. Defaults to 10.

**``AUTOTASK_RETRY_DELAY``**: Integer. Time in seconds autotask waits before executing a *@delayed_task* again in case an error has occured. Errors are unhandled exeptions. Defaults to 2.

**``AUTOTASK_CLEAN_INTERVALL``**: Integer. Time in seconds between database cleanup runs. After running a *@delayed_task* the result is stored for at least the given time to live (the decorator *ttl* parameter). After this period the entry will get removed by the next cleanup run to prevent the accumulation of outdated tasks in the database. Defaults to 600.





RELEASES.rst
Releases
--------

0.6
...

Major refactoring. Scaling decoupled from django processes.
New AUTOTASK_WORKERS setting.


0.5.3
.....

Bugfix: missing import added.
Some typos in admin interface.

0.5.2
.....

Timezone bugfix

0.5.1
.....

Bugfix: missing import added

0.5
...

PostgreSQL Requirement removed.
Works now with PostgreSQL and SQLite3.
(Should also work with MySQL and Oracle, but untested)

0.4
...

Rewrite of the cron scheduler

0.3.1
.....

Bugfix in settings import

0.3
...

Refactoring of settings import. If autotask is not active the decorators will  return the original function.

0.2.2
.....

AUTOTASK_IS_ACTIVE setting added to simplify installation.

0.2.1
.....

Bugfix in backport.

0.2
...

Backport for Python 2: runs now with Python >= 3.3 and 2.7

0.1.6
.....

worker main loop enhanced

0.1.5
.....

Bugfix: starting subprocess now with executable from settings.

0.1.4
.....

Bugfix in signal handling and executable settings added. Improved stability. Documentation extended.

0.1.0 - 0.1.3
.............

first public releases.



autotask/tasks.py
"""
Function-Decorators for async task-execution by a worker.
"""

import importlib
import pickle
from datetime import timedelta

from django.utils.timezone import now

from .conf import settings
from .cron import CronScheduler
from .models import (
    DONE,
    ERROR,
    TaskQueue,
)


class DelayedTask(object):
    """
    Gives access to a delayed function.
    """

    def __init__(self, pk):
        # pk of TaskQueue item in db
        self.pk = pk

    def _get_task(self):
        try:
            return TaskQueue.objects.get(pk=self.pk)
        except TaskQueue.DoesNotExist:
            return None

    @property
    def ready(self):
        """
        Returns True, False or None.
        In case of None the task is lost and no informations are
        available.
        """
        task = self._get_task()
        if task and not task.is_periodic:
            return task.status == DONE or task.status == ERROR
        return None

    @property
    def status(self):
        """
        Returns the processing status.
        """
        task = self._get_task()
        if task:
            return task.status
        return None

    @property
    def result(self):
        """
        Returns the result of the task.
        """
        task = self._get_task()
        if task:
            try:
                return pickle.loads(task.result)
            except EOFError:
                # if task.result is None this exception will raised
                pass
        return None

    @property
    def error_message(self):
        """
        Returns the error-message as string. If no error has occured
        returns an empty string.
        """
        task = self._get_task()
        if task:
            return task.error_message
        return ''


class DecoratorBase(object):
    """
    Common functionality for a decorator accepting arguments.
    """

    def __call__(self, function):
        """bind function-label to self.wrapper"""
        if not settings.AUTOTASK_IS_ACTIVE:
            # don't wrapp on inactive autotask
            return function
        self.module_name = function.__module__
        module = importlib.import_module(function.__module__)
        self.function_name = self.template.format(function.__name__)
        setattr(module, self.function_name, function)
        if not self.function_name.endswith('_delayed'):
            # a periodic task will never get called from the application
            # so the wrapper has to be called here:
            self.wrapper()
        return self.wrapper

    def wrapper(self, *args, **kwargs):
        """
        Gets called instead of function. Adds a TaskQueue item for the
        wrapped function and return a DelayedTask objects for accessing
        status-informations and optional results.
        """
        tq = TaskQueue()
        tq.arguments = pickle.dumps((args, kwargs))
        tq.module = self.module_name
        tq.function = self.function_name
        tq = self.configure(tq)
        if tq.is_periodic:
            if self.is_registered(tq):
                return None
        tq.save()
        dt = DelayedTask(tq.pk)
        return dt

    def is_registered(self, tq):
        """
        Returns a boolean whether a task is allready saved in the
        database. This is to prevent periodic tasks registered more than
        once at startup what will happen if the django-project is
        running with more than one process.
        """
        try:
            TaskQueue.objects.get(
                module=tq.module,
                function=tq.function,
                is_periodic=tq.is_periodic)
        except TaskQueue.DoesNotExist:
            return False
        return True


class delayed_task(DecoratorBase):  # noqa
    """
    Decorator to wrap a function for delayed execution by a separate
    worker process.
    :delay: worker waits n seconds before processing
    :retries: on error try to rerun the tasks n times
    :ttl: time to live: after processing the task will stay at least n
    seconds in the database i.e. for accessing the result.

        @delayed_task(optional arguments)
        def long_runner(*args, **kwargs)
        ...
        dt = long_runner()

    The returned object dt is of type DelayedTask
    """
    def __init__(self, delay=0, retries=0, ttl=300):
        self.ttl = timedelta(seconds=ttl)
        self.delay = timedelta(seconds=delay)
        self.retries = retries
        self.template = '{}_delayed'

    def configure(self, tq):
        tq.scheduled = now() + self.delay
        tq.retries = self.retries
        tq.ttl = self.ttl
        tq.is_periodic = False
        return tq


class periodic_task(DecoratorBase):  # noqa
    """
    Decorator for a periodic task running every given number of seconds
    Returns a DelayedTask object. Usage:

        @periodic_task()
        def some_function():
            ...

    Default period is 3600 seconds. If start_now is True the task will
    run as soon as possible and then periodically. Id start_now is False
    the task will be delayed by the given period before running
    periodically.
    """
    def __init__(self, seconds=3600, start_now=False):
        self.timedelta = timedelta(seconds=seconds)
        self.delay = timedelta() if start_now else self.timedelta
        self.template = '{}_periodic'

    def configure(self, tq):
        tq.scheduled = now() + self.delay
        tq.timedelta = self.timedelta
        tq.is_periodic = True
        return tq


class cron_task(DecoratorBase):  # noqa
    """
    Decorator for a periodic task running by a crontab-like schedule:

        @cron_task(optional arguments)
        def the_task()

    The arguments are:

    minutes: list of integers in the range 0-59 or None. [5, 15] will
    run 5 and 15 minutes after full hour. If None the task will run
    every minute.

    hours: list of integers in the range 0-23 or None. [5, 15] will run
    at 5 a.m. and 3 p.m. If None the task will run every hour.

    dow: list of integers in the range 0-6 or None. Describes the days
    of week when a task should run. Monday is 0 and Sunday is 6. [2, 4]
    will run every Wednesday and Friday. If None the task will run every
    day.

    month: list of integers in the range 1-12 or None. Describes the
    month of a year when a task should run. [4, 7] will run in April and
    July. If None the task will run every month.

    dom: list of integers in the range 1-31 or None. Describes the days
    of a month when a task should run. [4, 15] will run at the 4th and
    15th of every month. If None the task will run every day.


    Some examples:

    @cron_task(minutes=[15, 30], hours=[7, 20])
    Runs every day at 7:15, 7:30, 20:15, 20:30

    @cron_task(minutes=[30], hours=[7], dow=[0, 2])
    Runs every Monday and Wednesday at 7:30

    @cron_task(minutes=[30], hours=[7], dow=[0, 2], month=[4, 7])
    Runs every Monday and Wednesday at 7:30 in April and July

    @cron_task(minutes=[30], hours=[7], dom=[1, 15])
    Runs every 1st and 15th of a month at 7:30

    @cron_task(minutes=[30], hours=[7], dow=[0], dom=[1, 15])
    Runs every 1st and 15th of a month and at Mondays at 7:30

    @cron_task(minutes=[30], hours=[7], dow=[0], month=[4], dom=[1, 15])
    Runs at 1st and 15th and at Mondays at 7:30 but only in April

    If all arguments are None the task will run every minute, the same
    as @periodic_task(seconds=60).

    Instead of the separate arguments also a crontab with five patterns
    can be given (pattern order: minutes, hours, day of month, months,
    day of week):
    * * * * *       runs every minute (same as @periodic_task(seconds=60))
    30 7 * * 0,2    runs at 7:30 every Monday and Wednesday.

    """
    def __init__(self, minutes=None, hours=None,
                 dow=None, months=None, dom=None,
                 crontab=None):
        self.template = '{}_cron'
        self.cron_data = {
            'minutes': minutes,
            'hours': hours,
            'dow': dow,
            'months': months,
            'dom': dom,
            'crontab': crontab
        }

    def configure(self, tq):
        cs = CronScheduler(**self.cron_data)
        tq.scheduled = cs.get_next_schedule()
        tq.cron_data = pickle.dumps(self.cron_data)
        tq.is_periodic = True
        return tq
